core/src/main/java/org/apache/calcite/adapter/enumerable/EnumUtils.java

```diff
@@ -394,16 +392,6 @@ public static Expression convert(Expression operand, Type fromType,
               Expressions.unbox(operand, toBox),
               toBox));
     } else if (fromPrimitive != null && toBox != null) {
-      // E.g. from "int" to "Long".
-      // Generate Long.valueOf(x)
-      // Eliminate primitive casts like Long.valueOf((long) x)
-      if (operand instanceof UnaryExpression) {
-        UnaryExpression una = (UnaryExpression) operand;
-        if (una.nodeType == ExpressionType.Convert
-            || Primitive.of(una.getType()) == toBox) {
-          return Expressions.box(una.expression, toBox);
-        }
-      }
```

Comment:
If you check the following lines, there is
```
      // E.g., from "int" to "Byte".
      // Convert it first and generate "Byte.valueOf((byte)x)"
      // Because there is no method "Byte.valueOf(int)" in Byte
      return Expressions.box(
          Expressions.convert_(operand, toBox.primitiveClass),
          toBox);
```

So the case was aware of already, and it just seems like cases should skipped this optimization fail to do so.

Reply:
Yes, this line of code for general case is also added by me in [CALCITE-3414 ](https://github.com/apache/calcite/pull/1507).
But the optimization targets on the pattern like
`(byte)(int)->Byte`
In which the operand is `UnaryExpression` with `ExpressionType.Convert`